### PR TITLE
Dynamic discover ethclients in k8s cluster

### PR DESCRIPTION
**Pull request by tailingchen**
The first commit is refactor about dialing ethclients.
The second commit adds the k8s endpoint watch function to add new ethclients and removed unavailable ones.

### CODE REVIEW

File: multiclient/client.go [hunk 1]
```diff
@@ -726,3 +709,47 @@ func postToAll(ctx context.Context, fns []postFn) error {
...
+
+	for i := 0; i < len(urls); i++ {
+		dialed := <-dialCh
+		if dialed.client != nil {
+			mc.rpcClientMap.Replace(dialed.url, dialed.client)
+		}
+	}
+}
+
+func (mc *Client) retrydial() {
```
Comment (markya0616) [inline multiclient/client.go:765]: Should we have a wait group to wait the retry goroutine is done?

File: multiclient/k8s.go [hunk 2]
```diff
@@ -109,3 +95,186 @@ func createKubeClient(kubeconfig *KubeConfig) (clientset.Interface, error) {
...
+	defer s.mutex.Unlock()
+
+	// Add news urls
+	for _, url := range news {
+		if c := s.rpcClientMap.Get(url); c == nil {
+			s.rpcClientMap.Set(url, nil)
+		}
+	}
+
+	// Remove olds urls
```
Comment (markya0616) [inline multiclient/k8s.go]: Remove old urls

File: multiclient/k8s.go [hunk 3]
```diff
@@ -109,3 +95,186 @@ func createKubeClient(kubeconfig *KubeConfig) (clientset.Interface, error) {
...
+			}
+		}
+	}
+	return ethURLs
+}
+
+// findRemoved returns the string array represents the elements in olds array and removed
+// in news array.
+func findRemoved(olds, news []string) []string {
+	sort.Strings(olds)
```
Comment (markya0616) [inline multiclient/k8s.go]: remove sort

File: multiclient/k8s.go [hunk 4]
```diff
@@ -109,3 +95,186 @@ func createKubeClient(kubeconfig *KubeConfig) (clientset.Interface, error) {
...
+		return err
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	urls := s.endpoints[o.GetUID()]
+
+	for _, url := range urls {
+		s.rpcClientMap.Delete(url)
```
Comment (markya0616) [inline multiclient/k8s.go:233]: We should close client if the client is not nil.

File: multiclient/k8s.go [hunk 5]
```diff
@@ -109,3 +95,186 @@ func createKubeClient(kubeconfig *KubeConfig) (clientset.Interface, error) {
...
+	if err != nil {
+		return err
+	}
+	endpoint := obj.(*v1.Endpoints)
+	news := getEthURLsFromK8sEndpoint(endpoint, s.scheme)
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	// Add news urls
```
Comment (markya0616) [inline multiclient/k8s.go]: Add new urls

File: multiclient/client.go (hunk 1 above)
Comment (tailingchen) [reply multiclient/client.go:765]: Yes, we should.

File: multiclient/k8s.go (hunk 3 above)
Comment (tailingchen) [reply multiclient/k8s.go]: ok

**Review by markya0616 — APPROVED** [review]

**Pull request merged**